File: mobile/modules/background-downloads/ios/BackgroundDownloadsModule.swift
import ExpoModulesCore
import Foundation

struct BackgroundDownloadRequestRecord: Record {
  @Field var key: String = ""
  @Field var transferToken: String = ""
  @Field var accountScope: String = ""
  @Field var songId: String = ""
  @Field var scopes: [String] = []
  @Field var songJSON: String = ""
  @Field var audioURL: String = ""
  @Field var coverURL: String?
  @Field var lyricsURL: String?
  @Field var refreshURL: String = ""
  @Field var audioPath: String = ""
  @Field var coverPath: String?
  @Field var lyricsPath: String?
  @Field var priority: Double = 0.25
}

struct BackgroundDownloadReferenceRecord: Record {
  @Field var key: String = ""
  @Field var transferToken: String = ""
}

struct BackgroundDownloadAcknowledgementRecord: Record {
  @Field var key: String = ""
  @Field var transferToken: String = ""
  @Field var revision: Int64 = 0
}

public final class BackgroundDownloadsModule: Module {
  private var observer: NSObjectProtocol?

  public func definition() -> ModuleDefinition {
    Name("BackgroundDownloads")

    Events("stateChanged")

    OnCreate {
      self.observer = NotificationCenter.default.addObserver(
        forName: BackgroundDownloadCoordinator.stateChangedNotification,
        object: nil,
        queue: .main
      ) { [weak self] notification in
        guard
          let self,
          let payload = notification.userInfo as? [String: Any]
        else {
          return
        }
        self.sendEvent("stateChanged", payload)
      }
      BackgroundDownloadCoordinator.shared.restoreSession()
    }

    OnDestroy {
      if let observer = self.observer {
        NotificationCenter.default.removeObserver(observer)
      }
      self.observer = nil
    }

    AsyncFunction("enqueue") { (records: [BackgroundDownloadRequestRecord]) in
      let jobs = records.map {
        BackgroundDownloadRequest(
          key: $0.key,
          transferToken: $0.transferToken,
          accountScope: $0.accountScope,
          songId: $0.songId,
          scopes: $0.scopes,
          songJSON: $0.songJSON,
          audioURL: $0.audioURL,
          coverURL: $0.coverURL,
          lyricsURL: $0.lyricsURL,
          refreshURL: $0.refreshURL,
          audioPath: $0.audioPath,
          coverPath: $0.coverPath,
          lyricsPath: $0.lyricsPath,
          priority: $0.priority
        )
      }
      try BackgroundDownloadCoordinator.shared.enqueue(jobs)
    }

    AsyncFunction("cancel") { (records: [BackgroundDownloadReferenceRecord]) in
      try BackgroundDownloadCoordinator.shared.cancel(
        records.map {
          BackgroundDownloadReference(
            key: $0.key,
            transferToken: $0.transferToken
          )
        }
      )
    }

    AsyncFunction("cancelAccount") { (accountScope: String) in
      try BackgroundDownloadCoordinator.shared.cancelAccount(accountScope)
    }

    AsyncFunction("cancelAll") {
      try BackgroundDownloadCoordinator.shared.cancelAll()
    }

    AsyncFunction("snapshot") { (accountScope: String?) -> [[String: Any]] in
      BackgroundDownloadCoordinator.shared.snapshot(accountScope: accountScope)
    }

    AsyncFunction("acknowledge") {
      (records: [BackgroundDownloadAcknowledgementRecord]) in
      try BackgroundDownloadCoordinator.shared.acknowledge(
        records.map {
          BackgroundDownloadAcknowledgement(
            key: $0.key,
            transferToken: $0.transferToken,
            revision: $0.revision
          )
        }
      )
    }

    AsyncFunction("setActiveAccount") { (accountScope: String) in
      try BackgroundDownloadCoordinator.shared.setActiveAccount(accountScope)
    }
  }
}

File: mobile/modules/background-downloads/android/src/main/java/expo/modules/backgrounddownloads/BackgroundDownloadsModule.kt
package expo.modules.backgrounddownloads

import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class BackgroundDownloadsModule : Module() {
  private val mainHandler = Handler(Looper.getMainLooper())
  private var coordinator: BackgroundDownloadCoordinator? = null

  override fun definition() = ModuleDefinition {
    Name("BackgroundDownloads")

    Events("stateChanged")

    OnCreate {
      val context = appContext.reactContext?.applicationContext
        ?: return@OnCreate
      val instance = BackgroundDownloadCoordinator.get(context)
      coordinator = instance
      instance.listener = { payload ->
        mainHandler.post {
          sendEvent("stateChanged", payload)
        }
      }
      instance.restoreAndSchedule()
    }

    OnDestroy {
      coordinator?.listener = null
      coordinator = null
    }

    AsyncFunction("enqueue") { records: Array<BackgroundDownloadRequestRecord> ->
      requireCoordinator().enqueue(records.toList())
    }

    AsyncFunction("cancel") { records: Array<BackgroundDownloadReferenceRecord> ->
      requireCoordinator().cancel(records.toList())
    }

    AsyncFunction("cancelAccount") { accountScope: String ->
      requireCoordinator().cancelAccount(accountScope)
    }

    AsyncFunction("cancelAll") {
      requireCoordinator().cancelAll()
    }

    AsyncFunction("snapshot") { accountScope: String? ->
      requireCoordinator().snapshot(accountScope)
    }

    AsyncFunction("acknowledge") { records: Array<BackgroundDownloadAcknowledgementRecord> ->
      requireCoordinator().acknowledge(records.toList())
    }

    AsyncFunction("setActiveAccount") { accountScope: String ->
      requireCoordinator().setActiveAccount(accountScope)
    }
  }

  private fun requireCoordinator(): BackgroundDownloadCoordinator {
    val existing = coordinator
    if (existing != null) return existing
    val context = appContext.reactContext?.applicationContext
      ?: throw IllegalStateException("Background downloads are unavailable")
    return BackgroundDownloadCoordinator.get(context).also { coordinator = it }
  }
}
